docs/CONTRIBUTING.md

```diff
@@ -61,6 +61,24 @@ are squashed against the correct commit if necessary.
 
 ## Style Guide
 
+### Logging
+
+When possible, use log levels according to the following guidelines (in
+decreasing order of severity):
+
+- `error`: there is definitely a logic bug somewhere in the code, and someone
+  should look into it; it should not be possible to trigger an `error` log just
+  by manipulating inputs / configuration details
+- `warn`: something went wrong when processing an event or handling data; this
+  may be caused by invalid or malformed inputs
```

Comment:
2¢: I tend to think of the difference between `warn` and `error` as something roughly like:
- `error`: a logic bug has occurred, and processing beyond this point will definitely give the wrong result: an operation will fail, an error result will be returned, future processing may also go wrong because implicit invariants have been breached. Exclamation: "**uh oh!**".
 - `warn`: an indication that something unusual and *probably* wrong has happened. The situation may be technically valid, and the operation in progress can recover and return a sensible result, but there's a good chance that something wrong is afoot. Exclamation: "**huh, that's hinky?**".

In practice this style of separation often means that when there are a burst of `error`s in the logs, scrolling back to earlier `warn`ings gives clues as to root causes.

Reply:
I think we are on the same page (feel free to suggest better wording though of course), though the issue in my mind is whether we should ever log an `error` *from the runtime* if something *in the application* has a logic bug. My gut feel is that we should not, and we should let the application handle that (which may well decide to log it as `error`, as it would rightly be an error in the application, not the runtime).

I guess it's kind of confusing at the moment because we log both streams (runtime and application) in almost exactly the same way.

I added a paragraph below to try and clarify the distinction, PTAL.

Reply:
I tend in the other direction &ndash; I think it's entirely fine to log `error`s from the Runtime as a result of poor behaviour on the part of the application (e.g. passing invalid Wasm linear memory references, using invalid handles, passing blobs of data that are valid serialized protobufs).  That's in line with my wordings above, which distinguish `warn`/`error` much more based on consequences (what happens next) rather than root causes.

Downgrading anything that is application-triggerable to `warn` also seems like it would be a waste of an error level; on a quick count over 70% (45/62) of the current `error`s in the runtime are in that category.

Reply:
Just as a quick follow-up: often when I re-run tests and / or examples and I see actual `error`s logged, it makes me suspect that I broke something. Then I need to re-read all the actual error messages, and figure out that in fact they were pre-existing and "expected".

I am basically training my brain to ignore error messages, which seems a bad strategy long-term.

Reply:
For N-1 of the examples, I agree with you &ndash; we shouldn't see `error`s, and we should investigate any `error`s we do see.  This would include `error`s that indicate poor behaviour on behalf of the application &ndash; if (say) the `aggregator` example started emitting `error`s indicating invalid pointers or wrongly-serialized protobufs, then we should jump on that.

The fly in the ointment is the `abitest` example, which is deliberately and explicitly testing as many error paths as possible.  That's important for test coverage, but doesn't reflect normal use, and so its results don't help in terms of brain training. 

In terms of data, on a quick grep it looks like 41 of the 59 `error`s emitted by `run_examples` are from `abitest`; 16 of the remaining 18 `error`s could be removed by downgrading two lines of code to `warn`.

Reply:
#1211 *should* bring us down to just
```log
[2020-06-29T08:47:44Z ERROR oak_runtime::node::logger] Couldn't find the nearest Point Of Interest: Status { code: 5, message: "No open stations", details: [] }
...
[2020-06-29T08:48:34Z ERROR oak_runtime::node::logger] Couldn't find the nearest Point Of Interest: Status { code: 5, message: "No open stations", details: [] }
```
outside of `abitest`, and that would get separated by #1182 anyway.

Reply:
#1211 seems to bring us to zero `error`s from `run_examples` outside of `abitest` (and down to 33 from `abitest`).